common/xfer/pipes.go

```diff
@@ -27,7 +27,16 @@ type pipe struct {
 	onClose         func()
 }
 
-// NewPipe makes a new... pipe.
+// NewPipeFromEnds makes a new pipe specifying its ends
+func NewPipeFromEnds(local io.ReadWriter, remote io.ReadWriter) Pipe {
```

Comment:
After our conversation on Slack I now understand this. I need to think carefully about how to resolve this though. I am still not closely familiar with the full pipes though and I think this refactoring falls of out of the scope of this PR. I have created